src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -480,21 +502,27 @@ protected void writeHeapRecordPrologue() throws IOException {
             currentSegmentStart = fos.getChannel().position();
             // write dummy length of 0 and we'll fix it later.
             out.writeInt(0);
+        } else if (isCompression()) {
+            segmentOut.enterSegmentMode();
         }
     }
 
     @Override
     protected void writeHeapRecordEpilogue() throws IOException {
-        if (useSegmentedHeapDump) {
+        if (isCompression() == false && useSegmentedHeapDump) {
             out.flush();
             if ((fos.getChannel().position() - currentSegmentStart - 4L) >= HPROF_SEGMENTED_HEAP_DUMP_SEGMENT_SIZE) {
                 fillInHeapRecordLength();
                 currentSegmentStart = 0;
             }
+        } else if (isCompression()) {
+            segmentOut.exitSegmentMode();
         }
     }
 
     private void fillInHeapRecordLength() throws IOException {
+        // For compression, the length is written by CompressedSegmentOutputStream
+        if (isCompression()) return;
```

Comment:
I would expect to find logic in `CompressedSegmentOutputStream` that is similar to what is being skipped here in `fillInHeapRecordLength()`, but I don't. For example, I don't see a check for overflow that results in an exception like there is a few lines below here.